File: app/src/main/java/com/example/nfcdoorcard/utils/RootShell.java
package com.example.nfcdoorcard.utils;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.List;
import java.util.concurrent.TimeUnit;

public final class RootShell {
    private static final long TIMEOUT_SECONDS = 10;

    private RootShell() {}

    public static boolean run(List<String> commands) {
        if (commands == null || commands.isEmpty()) return true;
        StringBuilder script = new StringBuilder("set -e\n");
        for (String cmd : commands) {
            script.append(cmd).append('\n');
        }
        Result result = execute(script.toString());
        return result.success();
    }

    public static boolean run(String... commands) {
        return run(java.util.Arrays.asList(commands));
    }

    public static String runWithResult(String command) {
        Result result = execute(command);
        if (result.success()) return result.output();
        if (result.output().isEmpty()) {
            return "ERROR: exit=" + result.exitCode() + (result.timedOut() ? " timeout" : "");
        }
        return result.output();
    }

    public static Result execute(String command) {
        Process process = null;
        StringBuilder output = new StringBuilder();
        try {
            process = new ProcessBuilder("su", "-c", command)
                    .redirectErrorStream(true)
                    .start();

            Process finalProcess = process;
            Thread readerThread = new Thread(() -> {
                try (BufferedReader reader = new BufferedReader(
                        new InputStreamReader(finalProcess.getInputStream()))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        synchronized (output) {
                            output.append(line).append('\n');
                        }
                    }
                } catch (Exception ignored) {
                }
            }, "root-shell-reader");
            readerThread.start();

            boolean finished = process.waitFor(TIMEOUT_SECONDS, TimeUnit.SECONDS);
            if (!finished) {
                process.destroyForcibly();
                readerThread.join(500);
                return new Result(false, -1, trim(output), true);
            }

            readerThread.join(500);
            int exitCode = process.exitValue();
            return new Result(exitCode == 0, exitCode, trim(output), false);
        } catch (Exception e) {
            return new Result(false, -1, e.getClass().getSimpleName() + ": " + e.getMessage(), false);
        } finally {
            if (process != null) process.destroy();
        }
    }

    private static String trim(StringBuilder output) {
        synchronized (output) {
            return output.toString().trim();
        }
    }

    public record Result(boolean success, int exitCode, String output, boolean timedOut) {}
}

File: app/src/main/java/com/yagay/nfcdoorcard/system/RootShell.kt
package com.yagay.nfcdoorcard.system

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.widget.Toast
import java.util.concurrent.TimeUnit

class RootShell(context: Context) {
    companion object {
        private const val ROOT_OK_CACHE_MS = 60_000L
        private const val ROOT_FAILURE_CACHE_MS = 5_000L
    }

    private val appContext = context.applicationContext
    @Volatile private var rootAvailableCache: Boolean? = null
    @Volatile private var rootCacheCheckedAt: Long = 0L
    @Volatile private var lastRootToastAt: Long = 0L

    fun run(command: String, timeoutSeconds: Long = 20, maxChars: Int = 1_000_000, showToast: Boolean = true): String {
        if (!ensureRootAccess(showToast)) return "ROOT_UNAVAILABLE"
        return try {
            val process = ProcessBuilder("su", "-c", command).redirectErrorStream(true).start()
            val output = StringBuilder()
            val reader = Thread({
                runCatching {
                    process.inputStream.bufferedReader().useLines { lines ->
                        lines.forEach { line ->
                            if (output.length < maxChars) {
                                val remaining = maxChars - output.length
                                val piece = if (line.length + 1 <= remaining) line + "\n" else line.take(remaining)
                                output.append(piece)
                            }
                        }
                    }
                }
            }, "NfcDoorCard-RootReader").apply { isDaemon = true; start() }

            val finished = process.waitFor(timeoutSeconds, TimeUnit.SECONDS)
            if (!finished) {
                process.destroyForcibly()
                process.waitFor(2, TimeUnit.SECONDS)
            }
            reader.join(1500)
            if (!finished) output.appendLine("[timeout=${timeoutSeconds}s]")
            else output.appendLine("[exit=${process.exitValue()}]")
            output.toString()
        } catch (t: Throwable) {
            invalidateRootCache()
            if (showToast) notifyRootUnavailable()
            "ERROR ${t.javaClass.simpleName}: ${t.message}"
        }
    }

    fun invalidateRootCache() {
        rootAvailableCache = null
        rootCacheCheckedAt = 0L
    }

    private fun ensureRootAccess(showToast: Boolean): Boolean {
        val now = SystemClock.elapsedRealtime()
        rootAvailableCache?.let { cached ->
            val ttl = if (cached) ROOT_OK_CACHE_MS else ROOT_FAILURE_CACHE_MS
            if (now - rootCacheCheckedAt in 0 until ttl) {
                if (!cached && showToast) notifyRootUnavailable()
                return cached
            }
        }

        val ok = try {
            val process = ProcessBuilder("su", "-c", "id -u").redirectErrorStream(true).start()
            val finished = process.waitFor(4, TimeUnit.SECONDS)
            val output = if (finished) process.inputStream.bufferedReader().readText().trim() else ""
            if (!finished) process.destroyForcibly()
            finished && process.exitValue() == 0 && output.lineSequence().any { it.trim() == "0" }
        } catch (_: Throwable) {
            false
        }
        rootAvailableCache = ok
        rootCacheCheckedAt = now
        if (!ok && showToast) notifyRootUnavailable()
        return ok
    }

    private fun notifyRootUnavailable() {
        val now = SystemClock.elapsedRealtime()
        if (now - lastRootToastAt < 3000L) return
        lastRootToastAt = now
        Handler(Looper.getMainLooper()).post {
            Toast.makeText(appContext, "Root 获取失败，请在 Root 管理器中授予本应用权限", Toast.LENGTH_LONG).show()
        }
    }
}
